nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+
+        DataType chosenSimpleType;
+        if (nrOfCompatibleSimpleSubTypes == 0) {
+            chosenSimpleType = null;
+        } else if (nrOfCompatibleSimpleSubTypes == 1) {
+            chosenSimpleType = compatibleSimpleSubTypes.get(0);
+        } else {
+            chosenSimpleType = findMostSuitableType(value, compatibleSimpleSubTypes, Function.identity())
+                    .orElse(compatibleSimpleSubTypes.get(0));
+        }
+
+        return chosenSimpleType;
+    }
+
+    public static <T> Optional<T> findMostSuitableType(Object value, List<T> types, Function<T, DataType> dataTypeMapper) {
+        final Optional<T> mostSuitableType;
+
+        Optional<DataType> inferredDataTypeOptional = Optional.ofNullable(inferDataType(value, null))
+                .filter(dataType -> !dataType.getFieldType().equals(RecordFieldType.STRING));
+
+        if (value instanceof String) {
+            mostSuitableType = findMostSuitableTypeByStringValue((String) value, types, dataTypeMapper);
+        } else if (inferredDataTypeOptional.isPresent()) {
+            DataType inferredDataType = inferredDataTypeOptional.get();
+
+            Optional<T> inferredTypeOptional = types.stream()
+                    .filter(type -> dataTypeMapper.apply(type).equals(inferredDataType))
+                    .findFirst();
+
+            if (inferredTypeOptional.isPresent()) {
+                mostSuitableType = inferredTypeOptional;
+            } else {
+                Optional<T> widerAvailableTypeOptional = types.stream()
+                        .map(type -> getWiderType(dataTypeMapper.apply(type), inferredDataType).isPresent() ? type : null)
+                        .filter(Objects::nonNull)
+                        .findFirst();
+
+                if (widerAvailableTypeOptional.isPresent()) {
+                    mostSuitableType = widerAvailableTypeOptional;
+                } else {
+                    mostSuitableType = Optional.empty();
+                }
+            }
+        } else {
+            mostSuitableType = Optional.empty();
+        }
+
+        return mostSuitableType;
+    }
+
+    public static <T> Optional<T> findMostSuitableTypeByStringValue(String valueAsString, List<T> types, Function<T, DataType> dataTypeMapper) {
+        Optional<T> mostSuitableType = types.stream()
+                // Sorting based on the RecordFieldType enum ordering looks appropriate here as we want simpler types
+                //  first and the enum's ordering seems to reflect that
+                .sorted((type1, type2) -> {
+                            int comparison;
+
+                            RecordFieldType dataType1 = dataTypeMapper.apply(type1).getFieldType();
+                            RecordFieldType dataType2 = dataTypeMapper.apply(type2).getFieldType();
+
+                            // Moving TIMESTAMP at the front (at least it should precede DATE)
+                            if (dataType1 == RecordFieldType.TIMESTAMP) {
```

Comment:
This seems a bit convoluted. Does it not make sense to instead just move the TIMESTAMP before DATE in the enum, if this is what we want?

Reply:
Yeah, that would be the best. I wasn't sure if the current ordering to stay as it is was important for other reasons (didn't wan't to risk regression).
But if you think it's fine to change the ordering I can do that.